Considere a base de dados encontrada em Irisdat.xlsx. Daí, pede-se:
 a) Treinar um PMC que classifique observações de flores íris em 3 espécies (Setosa, Versicolor e Virginica) usando como
entradas as características SEPALLENGTH (SL), SEPALWIDTH (SW), PETALLENGTH (PL) e
PETALWIDTH (PW).

In [25]:
import pandas as pd

# URL do arquivo de dados no GitHub
url = 'https://raw.githubusercontent.com/Nethyaze/Redes_Neurais/main/Irisdat.dat'

# Carrega o arquivo de dados usando pandas com os separadores corretos
dados = pd.read_csv(url, delimiter='\t', decimal=',')

# Imprime as primeiras linhas para verificar se os dados foram carregados corretamente
print(dados.head())




   5,0  3,3  1,4  0,2
0  6.4  2.8  5.6  2.2
1  6.5  2.8  4.6  1.5
2  6.7  3.1  5.6  2.4
3  6.3  2.8  5.1  1.5
4  4.6  3.4  1.4  0.3


In [31]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

# URL do arquivo de dados no GitHub
url = 'https://raw.githubusercontent.com/Nethyaze/Redes_Neurais/main/Irisdat.dat'

# Carrega o arquivo de dados usando pandas com os separadores corretos
dados = pd.read_csv(url, delimiter='\t', decimal=',')

# Renomeia as colunas para os nomes corretos
dados.columns = ['SL', 'SW', 'PL', 'PW', 'TYPE']

# Separa as características (entradas) e os rótulos (saídas)
X = dados[['SL', 'SW', 'PL', 'PW']]
y = dados['TYPE']

# Divide os dados em conjunto de treino e teste (80% para treino, 20% para teste)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Inicializa o Perceptron Multicamadas (PMC) com uma única camada oculta de 100 neurônios
modelo_classificador = MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000, random_state=42)

# Treina o modelo usando os dados de treino
modelo_classificador.fit(X_train, y_train)

# Faz previsões usando o conjunto de teste
previsoes = modelo_classificador.predict(X_test)

# Calcula a acurácia do modelo
acuracia = accuracy_score(y_test, previsoes)
print('Acurácia do modelo de classificação: {:.2f}%'.format(acuracia * 100))

# Faz previsões usando o conjunto de teste
previsoes = modelo_classificador.predict(X_test)

# Imprime as previsões e os tipos reais
for previsto, real in zip(previsoes, y_test):
    print(f'Tipo previsto: {previsto}, Tipo real: {real}')

# Calcula a acurácia do modelo
acuracia = accuracy_score(y_test, previsoes)
print('Acurácia do modelo de classificação: {:.2f}%'.format(acuracia * 100))


Acurácia do modelo de classificação: 93.10%
Tipo previsto: SETOSA, Tipo real: SETOSA
Tipo previsto: VIRGINIC, Tipo real: VIRGINIC
Tipo previsto: VERSICOL, Tipo real: VERSICOL
Tipo previsto: SETOSA, Tipo real: SETOSA
Tipo previsto: SETOSA, Tipo real: SETOSA
Tipo previsto: VIRGINIC, Tipo real: VIRGINIC
Tipo previsto: SETOSA, Tipo real: SETOSA
Tipo previsto: VERSICOL, Tipo real: VERSICOL
Tipo previsto: VERSICOL, Tipo real: VERSICOL
Tipo previsto: VERSICOL, Tipo real: VERSICOL
Tipo previsto: SETOSA, Tipo real: SETOSA
Tipo previsto: VIRGINIC, Tipo real: VIRGINIC
Tipo previsto: VERSICOL, Tipo real: VERSICOL
Tipo previsto: VERSICOL, Tipo real: VERSICOL
Tipo previsto: VIRGINIC, Tipo real: VIRGINIC
Tipo previsto: SETOSA, Tipo real: SETOSA
Tipo previsto: SETOSA, Tipo real: SETOSA
Tipo previsto: VIRGINIC, Tipo real: VIRGINIC
Tipo previsto: VIRGINIC, Tipo real: VIRGINIC
Tipo previsto: VIRGINIC, Tipo real: VIRGINIC
Tipo previsto: VIRGINIC, Tipo real: VERSICOL
Tipo previsto: VIRGINIC, Tipo real: VIR

 b) Estime SL a partir de SW, PL, PW.

In [32]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# URL do arquivo de dados no GitHub
url = 'https://raw.githubusercontent.com/Nethyaze/Redes_Neurais/main/Irisdat.dat'

# Carrega o arquivo de dados usando pandas com os separadores corretos
dados = pd.read_csv(url, delimiter='\t', decimal=',')

# Renomeia as colunas para os nomes corretos
dados.columns = ['SL', 'SW', 'PL', 'PW', 'TYPE']

# Separa as características (entradas) e o alvo (SL) para a regressão
X_regressao = dados[['SW', 'PL', 'PW']]
y_regressao = dados['SL']

# Divide os dados em conjunto de treino e teste (80% para treino, 20% para teste)
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_regressao, y_regressao, test_size=0.2, random_state=42)

# Inicializa e treina o modelo de Regressão Linear
modelo_regressao = LinearRegression()
modelo_regressao.fit(X_train_reg, y_train_reg)

# Faz previsões usando o conjunto de teste
previsoes_reg = modelo_regressao.predict(X_test_reg)

# Calcula métricas de desempenho do modelo de regressão
mse = mean_squared_error(y_test_reg, previsoes_reg)
r2 = r2_score(y_test_reg, previsoes_reg)

print('Erro quadrático médio (MSE):', mse)
print('Coeficiente de determinação (R^2):', r2)


Erro quadrático médio (MSE): 0.09470979693558297
Coeficiente de determinação (R^2): 0.8753272301170404


O erro quadrático médio (MSE) de aproximadamente 0.0947 indica que o modelo tem uma precisão razoável na estimativa de SL com base nas outras características.

Além disso, o coeficiente de determinação (R²) de aproximadamente 0.8753 sugere que o modelo explica cerca de 87,53% da variabilidade em SL usando as características SW, PL e PW. Quanto mais próximo de 1 for o valor de R², melhor o modelo se ajusta aos dados.
